homeassistant/components/enphase_envoy/sensor.py

```diff
@@ -55,63 +64,103 @@
 )
 
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+async def async_setup_platform(
+    homeassistant, config, async_add_entities, discovery_info=None
+):
     """Set up the Enphase Envoy sensor."""
     ip_address = config[CONF_IP_ADDRESS]
     monitored_conditions = config[CONF_MONITORED_CONDITIONS]
     name = config[CONF_NAME]
     username = config[CONF_USERNAME]
     password = config[CONF_PASSWORD]
 
-    envoy_reader = EnvoyReader(ip_address, username, password)
+    if "inverters" in monitored_conditions:
+        envoy_reader = EnvoyReader(ip_address, username, password, inverters=True)
+    else:
+        envoy_reader = EnvoyReader(ip_address, username, password)
+
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        data = {}
+        async with async_timeout.timeout(30):
+            try:
+                await envoy_reader.getData()
+            except httpx.HTTPStatusError as err:
+                _LOGGER.error("Error retrieving data: %s", err)
```

Comment:
Why do we not raise `UpdateFailed` in this case?

Reply:
Inverters are the only monitored conditions that require authentication. My thought on this was to log an error only when authentication fails and do UpdateFailed for other errors. So that we continue to get data for the other conditions. I wasn't sure of how other integrations handle this. If others do an all or nothing approach I can change it from an error to UpdateFailed.

Reply:
If this is the only call that does I/O and it fails there won't be any new data. Am I missing something?

Reply:
Yeah, this is the only I/O call. The `getData()` call makes a batch retrieval from the device, then when the entity updates its' state we can still get the state of the data that was successful. I guess the side effect with the current implementation here is that every poll (60 seconds) an error is logged saying the Inverter retrieval received a 401 error.

If I go with only UpdateFailed no entities will be created upon setup and eventually will stop trying to setup up the platform after all the retries.

I guess the appropriate behavior is to just fail the setup completely and let the user either fix their `username`/`password` or remove `inverters` from `monitored_conditions` in the` configuration.yaml.`?

Reply:
We should fail platform set up if there's an auth error and let the user fix the auth credentials.

Reply:
I added this to check if the connection can be made to the device or not. If authentication fails, the setup fails and does not continue. If an HTTP error occurs then let HA retry and recover from the transient state.

https://github.com/home-assistant/core/blob/778a82959f27561fe716700f18a709d33f9a35c1/homeassistant/components/enphase_envoy/sensor.py#L82-L88